### PR TITLE
Filters refactor

### PR DESCRIPTION
**Pull request by ZeRego**
### Pre requirements:
- [x] I have followed (at least) the [PR section of the contributing guide](https://github.com/lightdash/lightdash/blob/main/.github/CONTRIBUTING.md#sending-a-pull-request).

### Reference:
Closes: #207 

### Description:

- reformat filters structure to handle complex logic
- refactor UI 
- flexible typing

Note: will refactor the sql builder once we introduce the OR operator and metric filters

### Preview:
<a href="https://www.loom.com/share/40bf29323ecd45b5a23e6d19c8c906e4">
    <p>Lightdash - filters - Watch Video</p>
    
  </a>

### Scope of the changes (select all that apply):
- [x] Frontend  
- [x] Backend  
- [x] Common  
- [ ] E2E tests  
- [ ] Docs  
- [ ] CI/CD

### CODE REVIEW

File: packages/frontend/src/components/common/Filters/index.tsx [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+                    flexDirection: 'column',
+                    gap: '5px',
+                }}
+            >
+                {rules.length > 0 && (
+                    <div
+                        style={{ display: 'inline-flex', alignItems: 'center' }}
+                    >
+                        <p style={{ margin: 0, marginRight: 10 }}>
+                            Operator: <Tag minimal>AND</Tag>
```
Comment (owlas) [inline packages/frontend/src/components/common/Filters/index.tsx:96]: This feels irrelevant when I've only got one filter:
Screenshot 2022-01-31 at 15 59 29

File: packages/frontend/src/components/common/Filters/FilterRuleForm.tsx [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+                        settings: undefined,
+                        values: undefined,
+                    });
+                }
+            }
+        },
+        [activeField, fields, filterRule, onChange],
+    );
+
+    return (
```
Comment (owlas) [inline packages/frontend/src/components/common/Filters/FilterRuleForm.tsx:64]: Is this intentionally full width? 

Screenshot 2022-01-31 at 16 00 46

File: packages/frontend/src/components/common/Filters/FilterRuleForm.tsx [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+                        onChange={onChange}
+                    />
+                </>
+            ) : (
+                <span style={{ width: '100%', color: Colors.GRAY1 }}>
+                    Tried to reference field with unknown id:{' '}
+                    {filterRule.target.fieldId}
+                </span>
+            )}
+            <Button minimal icon="cross" intent="danger" onClick={onDelete} />
```
Comment (owlas) [inline packages/frontend/src/components/common/Filters/FilterRuleForm.tsx:98]: I don't think we need the coloured intent because it looks like a primary action

File: packages/backend/src/queryBuilder.ts [hunk 4]
```diff
@@ -1,125 +1,133 @@
...
 const renderNumberFilterSql = (
     dimensionSql: string,
-    filter: NumberFilter,
+    filter: FilterRule,
 ): string => {
     const filterType = filter.operator;
     switch (filter.operator) {
-        case 'equals':
-            return filter.values.length === 0
+        case FilterOperator.EQUALS:
```
Comment (owlas) [inline packages/backend/src/queryBuilder.ts:58]: I was able to reach this with a string argument:
Screenshot 2022-01-31 at 16 02 39
Screenshot 2022-01-31 at 16 02 35

File: packages/frontend/src/components/common/Filters/index.tsx (hunk 1 above)
Comment (owlas) [reply packages/frontend/src/components/common/Filters/index.tsx:96]: Screenshot 2022-01-31 at 16 15 11

Alternative could be to use a left margin for this information

**Review by owlas — CHANGES_REQUESTED** [review]

File: packages/frontend/src/components/common/Filters/FilterRuleForm.tsx (hunk 2 above)
Comment (ZeRego) [reply packages/frontend/src/components/common/Filters/FilterRuleForm.tsx:64]: Yes, so all rows independently of the input feel the same.

File: packages/frontend/src/components/common/Filters/index.tsx (hunk 1 above)
Comment (ZeRego) [reply packages/frontend/src/components/common/Filters/index.tsx:96]: I was trying to follow the same UI as this [library](https://react-querybuilder.js.org/react-querybuilder/#addRuleToNewGroups=false&autoSelectField=true&disabled=false&enableDragAndDrop=false&independentCombinators=false&resetOnFieldChange=true&resetOnOperatorChange=false&showCloneButtons=false&showCombinatorsBetweenRules=false&showNotToggle=false&validateQuery=false) 
In my next PR I'm going to make that label a dropdown to change between AND and OR. So it makes sense to only have it in 1 place. Open to suggestions.

File: packages/frontend/src/components/common/Filters/index.tsx [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+};
+
+const FiltersForm: FC<Props> = ({ fields, filters, setFilters }) => {
+    const rules: Array<FilterRule> = getTotalFilterRules(filters);
+
+    const onDeleteFilterRule = useCallback(
+        (index: number) => {
+            setFilters({
+                dimensions: {
+                    id: uuidv4(),
```
[the remark this answers is not in the document]
Comment (ZeRego) [reply packages/frontend/src/components/common/Filters/index.tsx:28]: The filter group id isn't used currently so i coded it like this for simplicity.

File: packages/frontend/src/components/common/Filters/index.tsx [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+    );
+
+    const onChangeFilterRule = useCallback(
+        (index: number, filterRule: FilterRule) => {
+            setFilters({
+                dimensions: {
+                    id: uuidv4(),
+                    and: [
+                        ...rules.slice(0, index),
+                        filterRule,
```
[the remark this answers is not in the document]
Comment (ZeRego) [reply packages/frontend/src/components/common/Filters/index.tsx:43]: The main use for id is so react knows the items order when rendering a list. By keeping the ids React knows the order hasn't changed.

File: packages/frontend/src/components/common/Filters/index.tsx (hunk 1 above)
Comment (owlas) [reply packages/frontend/src/components/common/Filters/index.tsx:96]: Screenshot 2022-01-31 at 16 57 35

I love this for inspiration! 

I think the placement at the side makes sense (vs. the top). Also in this example it avoids the problem of having many "AND"s if you want a single place to change the type

Comment (owlas) [reply packages/frontend/src/components/common/Filters/index.tsx:96]: But this is for another PR right?

Comment (ZeRego) [reply packages/frontend/src/components/common/Filters/index.tsx:96]: cc: @PriPatel

File: packages/frontend/src/components/Explorer.tsx [hunk 7]
```diff
@@ -215,7 +222,11 @@ export const Explorer: FC<Props> = ({ savedQueryUuid }) => {
                     ) : null}
                 </div>
                 <Collapse isOpen={filterIsOpen}>
-                    <FiltersForm />
+                    <FiltersForm
+                        fields={filterableDimensions}
```
[the remark this answers is not in the document]
Comment (owlas) [reply packages/frontend/src/components/Explorer.tsx:226]: @ZeRego what is the preferred behaviour? That the sidebar is selectable for `Created` or that `Created` is not a valid filter?

File: packages/frontend/src/components/common/Filters/FieldAutoComplete.tsx [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+};
+
+type Props = {
+    activeField: Field;
+    fields: Field[];
+    onChange: (value: string) => void;
+};
+
+const FieldAutoComplete: FC<Props> = ({ activeField, fields, onChange }) => (
+    <FieldSuggest
```
[the remark this answers is not in the document]
Comment (owlas) [reply packages/frontend/src/components/common/Filters/FieldAutoComplete.tsx:34]: @ZeRego the minimal popover now hides the autocomplete sometimes (not always)

Screenshot 2022-02-01 at 09 13 30

File: packages/frontend/src/components/TableTree.tsx [hunk 9]
```diff
@@ -146,7 +150,7 @@ const NodeItemButtons: FC<{
             />,
         );
     }
-    if (node.fieldType === FieldType.DIMENSION) {
+    if (node.fieldType === FieldType.DIMENSION && isFilterableDimension(node)) {
```
Comment (owlas) [inline packages/frontend/src/components/TableTree.tsx:153]: Great catch 🎉

**Review by owlas — CHANGES_REQUESTED** [review]
Couple of UI questions

File: packages/frontend/src/components/Explorer.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (ZeRego) [reply packages/frontend/src/components/Explorer.tsx:226]: This is a bug that exists in the app currently. Opened a [bug](https://github.com/lightdash/lightdash/issues/1266) for that

File: packages/frontend/src/components/common/Filters/configs.ts [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+            FilterOperator.LESS_THAN,
+            FilterOperator.GREATER_THAN,
+        ]),
+        inputs: DefaultFilterInputs,
+    },
+    [DimensionType.DATE]: {
+        operatorOptions: timeFilterOptions,
+        inputs: DateFilterInputs,
+    },
+    [DimensionType.TIMESTAMP]: {
```
[the remark this answers is not in the document]
Comment (ZeRego) [reply packages/frontend/src/components/common/Filters/configs.ts:79]: The `DateFilterInputs` component will be responsible for showing the correct inputs based on the field and operation.

File: packages/frontend/src/components/common/Filters/FieldAutoComplete.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (ZeRego) [reply packages/frontend/src/components/common/Filters/FieldAutoComplete.tsx:34]: Blueprint moves the popup like this when the dropdown list is to close to the bottom on the window. Doesn't matter the `minimal` and `position` props. Can't seem to find a fix for this. :/

[the remark this answers is not in the document]
Comment (owlas) [reply packages/frontend/src/components/common/Filters/FieldAutoComplete.tsx:34]: 💩

**Review by owlas — APPROVED** [review]
